src/Access/LDAPAccessStorage.cpp

```diff
@@ -0,0 +1,182 @@
+#include <Access/LDAPAccessStorage.h>
+#include <Access/User.h>
+#include <Common/Exception.h>
+#include <common/logger_useful.h>
+#include <ext/scope_guard.h>
+#include <Poco/Util/AbstractConfiguration.h>
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+
+LDAPAccessStorage::LDAPAccessStorage() : IAccessStorage("ldap")
+{
+}
+
+
+void LDAPAccessStorage::setConfiguration(const Poco::Util::AbstractConfiguration & config, IAccessStorage * top_enclosing_storage_)
```

Comment:
AccessControlManager could just pass a pointer to UsersConfigAccessStorage. Or you can implement getting the name of the storage where you're looking for the user template from the main config along with the setting `user_template`. The logic with checking `helper_lookup_in_progress` looks too complex.

Reply:
> AccessControlManager could just pass a pointer to UsersConfigAccessStorage

Reply:
The logic is ugly, but at least it is well localized here...

Reply:
> > AccessControlManager could just pass a pointer to UsersConfigAccessStorage
> 
> I need to look potentially everywhere, not just in `UsersConfigAccessStorage`.

I suppose the admin who has written `user_template` in the main config should know where the `user_template` is stored so he can write the name of the access storage containing that template user in the main config as well.

Reply:
What if the user is changing its "location" from `users.xml` to local directory and so on? I went on the path of least assumptions and most flexibility here.

Reply:
I don't think it's what we should care about. I mean an user can't do that, only the admin can, and it looks like a very rare situation.

Reply:
Right, but that is easily achieved here currently. And I don't see a way to change the current approach that wouldn't be several times more complicated than what we have here.

Reply:
Ok, but please implement it in easier way. You can pass a shared pointer to the list of access storages to your LDAPAccessStorage and just iterate through that list skipping LDAPAccessStorage itself.

Reply:
Changed.